src/client/java/teammates/client/scripts/sql/DataMigrationForTeamSql.java

```diff
@@ -0,0 +1,94 @@
+package teammates.client.scripts.sql;
+
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.stream.Collectors;
+
+import com.googlecode.objectify.cmd.Query;
+
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.JoinType;
+import jakarta.persistence.criteria.Root;
+
+import teammates.common.util.HibernateUtil;
+import teammates.storage.entity.Course;
+import teammates.storage.entity.CourseStudent;
+import teammates.storage.sqlentity.Section;
+import teammates.storage.sqlentity.Team;
+
+/**
+ * Data migration class for team entity.
+ */
+@SuppressWarnings("PMD")
+public class DataMigrationForTeamSql extends
+        DataMigrationEntitiesBaseScriptSql<Course, teammates.storage.sqlentity.Team> {
+
+    public static void main(String[] args) {
+        new DataMigrationForTeamSql().doOperationRemotely();
+    }
+
+    @Override
+    protected Query<Course> getFilterQuery() {
+        return ofy().load().type(Course.class);
+    }
+
+    @Override
+    protected boolean isPreview() {
+        return false;
+    }
+
+    /*
+     * Sets the migration criteria used in isMigrationNeeded.
+     */
+    @Override
+    protected void setMigrationCriteria() {
+        // No migration criteria currently needed.
+    }
+
+    @Override
+    protected boolean isMigrationNeeded(Course entity) {
+        return true;
+    }
+
+    @Override
+    protected void migrateEntity(Course oldCourse) throws Exception {
+        HibernateUtil.beginTransaction();
+        teammates.storage.sqlentity.Course newCourse = getNewCourse(oldCourse.getUniqueId());
+        HibernateUtil.commitTransaction();
+
+        getTeamNameToSectionNameMap(oldCourse)
+                .forEach((teamName, sectionName) -> createTeam(teamName, sectionName, newCourse.getSections()));
+    }
+
+    private teammates.storage.sqlentity.Course getNewCourse(String courseId) {
+        CriteriaBuilder cb = HibernateUtil.getCriteriaBuilder();
+        CriteriaQuery<teammates.storage.sqlentity.Course> cr = cb.createQuery(teammates.storage.sqlentity.Course.class);
+        Root<teammates.storage.sqlentity.Course> courseRoot = cr.from(teammates.storage.sqlentity.Course.class);
+        courseRoot.fetch("sections", JoinType.LEFT); // Fetch sections to avoid lazy-loading
```

Comment:
Does this improve performance? I would think since this is only 1 course, there would be no difference

If there is, is there a difference putting the fetch here and after you narrowed down the rows in line 71?

Reply:
Without `courseRoot.fetch(...)`, we are unable to call `s.getName()` outside of the transaction as it will result in an error - `could not initialize proxy - no Session`.

I timed a few runs with various combinations, with the same 10k teams:

| | run 1 | run 2 | run 3|
|----------|----------|----------|----------|
| query fetch (current) | 1m 17s | 1m 16s | 1m 17s |
| query fetch after line 71 | 1m 18s | 1m 17s | 1m 15s |
| query no fetch with map | 1m 20s | 1m 21s | 1m 19s |
| query fetch with map | 1m 11s | 1m 13s | 1m 12s |
| getRef with map | 1m 20s | 1m 21s | 1m 22s |
| get with map | 1m 19s | 1m 18s | 1m 18s |

Looks like moving the fetch to after line 71 might not have a large impact, but the most optimised setup is a query fetch with a `SectionNameToSectionMap`.

Reply:
I see!

Also when I mentioned removing the fetch, I thought you would move the commitTransaction to the end of migrateEntity. I think this could make each migrateEntity only use one transaction (I guess this assumed that all teams can be stored in memory which I think its quite probably)

Could you see if theres any performance benefit if you - move commit transaction to the end of MigrateEntity and remove the fetch section?

Reply:
The performance of moving `commitTransaction` to the end with no other changes is similar to `query no fetch with map` above. 

From the sql logs, using fetch results in one select instead of two, which seems to provide performance benefits.
```
Hibernate: select ... from courses c1_0 left join sections s1_0 on c1_0.id=s1_0.course_id where c1_0.id=?
...
```
vs
```
Hibernate: select ... from courses c1_0 where c1_0.id=?
Hibernate: select ... from sections s1_0 where s1_0.course_id=?
...
```

Using `persist` instead of `saveEntityDefered` reduces migration time to ~47s. 

Using `stream().filter(...).findfirst()` to retrieve `Section` from `sectionName` also seems to save a couple seconds, however this is with a small `MAX_SECTION_PER_COURSE` of 10, so I'll just leave the map in for now.